Name: Rezoan Hossain
Week 7 Assignment #5

In [1]:
import pandas as pd

# Load the movie ratings file
movies = pd.read_csv("movie_ratings.csv")

# Show the data
movies

,User,American Sniper,Edge of Tomorrow,Groundhog Day,Jurassic World,Lost in Translation,Lucy
0,John,5.0,4.0,3,NaN,NaN,4.0
1,Logan,4.0,NaN,3,3.0,NaN,NaN
2,Modesto,NaN,NaN,4,NaN,4.0,4.0
3,Malcolm,NaN,NaN,2,NaN,4.0,NaN
4,Maurice,5.0,4.0,4,2.0,3.0,3.0


In [2]:
# Average rating by user
# In this part, I am finding the average movie rating for each user.

user_avg = movies.set_index("User").mean(axis=1)

user_avg

User
John       4.000000
Logan      3.333333
Modesto    4.000000
Malcolm    3.000000
Maurice    3.500000
dtype: float64

In [3]:
# Average rating by movie
# In this part, I am finding the average rating for each movie.

movie_avg = movies.drop("User", axis=1).mean(axis=0)

movie_avg

American Sniper        4.666667
Edge of Tomorrow       4.000000
Groundhog Day          3.200000
Jurassic World         2.500000
Lost in Translation    3.666667
Lucy                   3.666667
dtype: float64

In [4]:
# Normalized ratings by user
# In this part, I am normalizing the ratings by subtracting
# each user's average rating from that user's movie ratings.

ratings_only = movies.set_index("User")

normalized = ratings_only.sub(user_avg, axis=0)

normalized

,American Sniper,Edge of Tomorrow,Groundhog Day,Jurassic World,Lost in Translation,Lucy
User,,,,,,
John,1.000000,0.0,-1.000000,NaN,NaN,0.0
Logan,0.666667,NaN,-0.333333,-0.333333,NaN,NaN
Modesto,NaN,NaN,0.000000,NaN,0.0,0.0
Malcolm,NaN,NaN,-1.000000,NaN,1.0,NaN
Maurice,1.500000,0.5,0.500000,-1.500000,-0.5,-0.5


In [5]:
# Average normalized rating by user
# In this part, I am finding the average normalized rating for each user.

normalized_user_avg = normalized.mean(axis=1)

normalized_user_avg

User
John       0.000000e+00
Logan     -1.480297e-16
Modesto    0.000000e+00
Malcolm    0.000000e+00
Maurice    0.000000e+00
dtype: float64

In [6]:
# Average normalized rating by movie
# In this part, I am finding the average normalized rating for each movie.

normalized_movie_avg = normalized.mean(axis=0)

normalized_movie_avg

American Sniper        1.055556
Edge of Tomorrow       0.250000
Groundhog Day         -0.366667
Jurassic World        -0.916667
Lost in Translation    0.166667
Lucy                  -0.166667
dtype: float64

# Conclusion

In this assignment, I created a movie ratings DataFrame for 5 users and 6 movies. Then I calculated the average rating by user and the average rating by movie. Based on the original averages, **American Sniper** had the highest average rating at **4.67**, while **Jurassic World** had the lowest average rating at **2.50**.

Next, I normalized the ratings by subtracting each user's average rating from that user's movie ratings. After normalization, the average normalized rating for each user was about **0**, which is expected. This shows that the normalization worked correctly.

The normalized movie averages showed that **American Sniper** still had the highest adjusted rating at about **1.06**, while **Jurassic World** had the lowest adjusted rating at about **-0.92**. This means American Sniper was rated above users’ personal averages, while Jurassic World was rated below users’ personal averages.

One advantage of normalized ratings is that they make movie comparisons more fair because they adjust for users who rate movies higher or lower than other users. This helps reduce user bias in the ratings. One disadvantage is that normalized ratings can be harder to understand than the original ratings, because the values are no longer the actual 1 to 5 movie scores. Another disadvantage is that normalization may remove some information about the users’ true rating style.

Overall, normalized ratings are useful when comparing movies across different users, but the original ratings are still easier to read and understand.